frontend/src/data/auth.ts

```diff
@@ -0,0 +1,246 @@
+import { useState } from "react";
+import * as jose from "jose";
+
+export enum UserClass {
+  client = "client",
+  staff = "staff",
+  manager = "manager",
+}
+
+// Parsed token payload
+export interface Token {
+  name: string;
+  iat?: Date;
+  sub: UserClass;
+  exp?: Date;
+}
+
+// Store state
+interface State {
+  // raw jwt
+  jwt: string;
+  // parsed jwt
+  token?: Token;
+}
+
+// Store actions
+interface Actions {
+  // set token on successful login
+  setToken: (jwt: string, token: Token) => void;
+  // clear, effectively logout
+  clear: () => void;
+}
+
+// Main store. For use within this file only. State in this store is shared across all components.
+const useAuthStore = create(
+  immer<State & Actions>((set) => ({
+    jwt: "",
+
+    setToken: (jwt, token) => {
+      set((state) => {
+        state.jwt = jwt;
+        state.token = token;
+      });
+    },
+
+    clear: () => {
+      set((state) => {
+        state.jwt = "";
+        state.token = undefined;
+      });
+    },
+  }))
+);
+
+export interface UseAuthResult {
+  name: string;
+  userClass: UserClass;
+  jwt: string;
+}
+
+// Custom AuthenticationErrror
+export class AuthenticationErrror extends Error {
+  constructor(message: string) {
+    super(message);
+    this.name = "AuthenticationError";
+  }
+}
+
+/**
+ * React hook for consumer of auth states
+ * This one is intended for components that only read these information
+ * Returns undefined if not logged in
+ */
+export const useAuth = (): UseAuthResult | undefined => {
+  const [jwt, token] = useAuthStore((state) => [state.jwt, state.token]);
+
+  if (token) {
+    return {
+      name: token.name,
+      userClass: token.sub,
+      jwt,
+    };
+  }
+};
+
+export interface UseAuthManangementResult {
+  // when true, show a loading indicator and block user from clicking login button
+  logginIn: boolean;
+  loginWithPassword: (username: string, password: string) => Promise<void>;
+  loginWithQR: (data: string) => Promise<void>;
+  logout: () => void;
+  // show error if set. no error if undefined
+  error?: Error;
+  clearError: () => void;
+}
+
+/**
+ * React hook for consumer of auth states
+ * This one is intended for components that manages logging in/out
+ */
+export const useAuthManagement = (): UseAuthManangementResult => {
+  // lock the UI
+  const [loggingIn, setLoggingIn] = useState(false);
+  // error to be displayed on the UI
+  const [error, setError] = useState<Error>();
+  // access authstore
+  const authStore = useAuthStore((state) => state);
+  // query public key from server
+  const {
+    error: pkQueryError,
+    data: pkQueryData,
+    isLoading: pkQueryLoading,
+  } = useQuery({
+    refetchOnWindowFocus: false,
+    cacheTime: PUBLIC_KEY_CACHE_TIME,
+    queryKey: ["repoData"],
+    queryFn: async () => {
+      const res = await fetch(API_ROOT + "/auth/publicKey");
+      if (res.ok) {
+        return jose.importSPKI(await res.text(), "ES256");
```

Comment:
Response isn't publicKey as plain text, but a json object with public key as `publicKey` key(lol it sounds funny)

Reply:
done